### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by pienaarp** [wkhtmltopdf/wkhtmltopdf#2343]
Hello,

I am using Wkhtmltopdf which is very nice, thank you. The output is very clear.

However there seems to be a memory leak, I think it may be caused by calls to 'wkhtmltopdf_convert', and for me the amount varies between 20k - 60k. 

Are you aware of this ? Do you have any advice to avoid ? I am definitely calling wkhtmltopdf_destroy_converter.  

Many thanks,

-Phil

**Comment by ashkulz** [comment]
There are memory leaks, but it's hard to fix out exactly where they are. They may be present in the Qt/WebKit bindings itself (which should hopefully be fixed with the 0.13 upgrade) or in the custom Qt patches (which I'll review for 0.13). The simplest solution is to use a separate process and periodically restart it if the leaks go very high.

**Comment by pienaarp** [comment]
Hi Ashish, thanks for the response.

Another post from Sept 2014 mentions the deleteLater() method, do you not
think this is the cause ? I didnt see any follow ups to that post ...

Phil

**Comment by ashkulz** [comment]
Yeah, but those were necessary for avoiding some crashes according to @pruiz. I've not really explored that in detail -- when you build in debug mode, it prints the leaks in WebKit after conversion completes. That might be a good start to identify where the problem is, but never got time to do that :frowning:

**Comment by pienaarp** [comment]
I understand ... thanks anyway, for now Im going to take your advice and
run out of process.

Cheers !

Phil

**Comment by ashkulz** [comment]
Can you try if 9fa4695ace411ea6e27e68c790c62f1e63646017 improves things? See the commit message for more details.

**Comment by ashkulz** [comment]
@pienaarp: were you able to try the above commit?

**Comment by SAYellow** [comment]
Hi Ashish, 

Is this still an open issue? As I am experiencing the same memory leaks with v 12.4.0...

Thank you for a gr8 app ;) 

Dirk

**Comment by alexander-nesterenko-kas** [comment]
Built **wkhtmltopdf 0.12.2.4** with the flag **-debug** and executed conversion HTML file into PDF under **valgrind-3.11.0**.

The command line:
```
valgrind --tool=memcheck --leak-check=full --show-leak-kinds=all --track-origins=yes --log-file=/home/alex/Desktop/sample-html-to-pdf.dbg.log ./wkhtmltopdf -q /home/alex/Desktop/sample.html /home/alex/Desktop/sample2.pdf 
```
The summary of report:
```
==26162== LEAK SUMMARY:
==26162==    definitely lost: 8,248 bytes in 124 blocks
==26162==    indirectly lost: 29,112 bytes in 418 blocks
==26162==      possibly lost: 336 bytes in 2 blocks
==26162==    still reachable: 294,108 bytes in 4,759 blocks
==26162==         suppressed: 0 bytes in 0 blocks
==26162== 
==26162== For counts of detected and suppressed errors, rerun with: -v
==26162== ERROR SUMMARY: 131 errors from 131 contexts (suppressed: 0 from 0)
```

[files.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/4276382/files.zip)
